### PR TITLE
support specifying time field in TimeBasedPartitioner

### PR DESCRIPTION
**Pull request by simshi**
Try to solve #98

### CODE REVIEW

**Comment by simshi** [comment]
Hi robot, 
[clabot:check]

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 1]
```diff
@@ -190,6 +190,12 @@
   public static final String PARTITION_FIELD_NAME_DEFAULT = "";
   public static final String PARTITION_FIELD_NAME_DISPLAY = "Partition Field Name";
 
+  public static final String PARTITION_TIME_FIELD_NAME_CONFIG = "partition.timefield.name";
```
Comment (ewencp) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Normally we'd separate each word in the config name with a `.` (as you separate each in the variable name with a `_`), so `"partition.time.field.name"` would probably be better here.

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 2]
```diff
@@ -190,6 +190,12 @@
...
+  private static final String PARTITION_TIME_FIELD_NAME_DOC =
+          "The timestamp field name of SinkRecord when TimeBasedPartitioner is used, "
+          + "current time will be used on empty";
+  public static final String PARTITION_TIME_FIELD_NAME_DEFAULT = "";
```
Comment (ewencp) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Can we leave it without a default instead? This is more natural and allows for better error checking as you can tell the user if they specify an empty (and therefore invalid) setting (which they may do, e.g., accidentally and need to clean up their config).

File: src/main/java/io/confluent/connect/hdfs/partitioner/TimeBasedPartitioner.java [hunk 3]
```diff
@@ -86,14 +88,23 @@ public void configure(Map<String, Object> config) {
...
   }
 
   @Override
   public String encodePartition(SinkRecord sinkRecord) {
-    long timestamp = System.currentTimeMillis();
+    long timestamp;
+    if (timeFieldName.equals("")) {
+      timestamp = System.currentTimeMillis();
+    } else {
+      Struct struct = (Struct) sinkRecord.value();
```
Comment (ewencp) [inline src/main/java/io/confluent/connect/hdfs/partitioner/TimeBasedPartitioner.java]: Probably want to validate the type before casting so you can throw the appropriate `ConnectException` if it isn't a `Struct`. Same deal for checking whether the field exists and if it's type is correct.

File: src/main/java/io/confluent/connect/hdfs/partitioner/TimeBasedPartitioner.java [hunk 4]
```diff
@@ -86,14 +88,23 @@ public void configure(Map<String, Object> config) {
...
+      timestamp = (long) struct.get(timeFieldName);
```
Comment (ewencp) [inline src/main/java/io/confluent/connect/hdfs/partitioner/TimeBasedPartitioner.java]: There are also logical types (e.g. `Timestamp`) that you might want to support here.

**Comment by simshi** [comment]
Thank you, agreed on most of your comments except I don't know how to handle logical types like `Timestamp`. Support this feature in another PR? :smile:

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 2 above)
Comment (simshi) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: My first attempt on eliminating default value failed, looked into `ConfigDef.parse()`, can't find a way to express "this config key is required only when partitioner.class is `TimeBasedPartitioner`". Suggestions are welcomed.

**Comment by jocelyndrean** [comment]
Hey ! Thanks @simshi for your PR.  No news on it since Aug 30 ? No merge ? :/

**Comment by simshi** [comment]
I'm not working on this, discarded, please roll your own version 😄

**Pull request closed without merging**